src/Core/ScreenshotCache.vala

```diff
@@ -1,141 +1,152 @@
-    private void delete_oldest_files (int screenshot_usage) {
-        string? name = null;
-        Dir dir;
-        if (read_dir (out dir, screenshot_path) == -1) {
+    private async void delete_oldest_files (uint64 screenshot_usage) {
+        var file_list = new Gee.ArrayList<GLib.FileInfo> ();
+
+        FileEnumerator enumerator;
+        try {
+            enumerator = yield screenshot_folder.enumerate_children_async (
+                string.join (",", "standard::*", GLib.FileAttribute.TIME_CHANGED),
+                FileQueryInfoFlags.NONE
+            );
+        } catch (Error e) {
+            warning ("Unable to create enumerator to delete cached screenshots: %s", e.message);
             return;
         }
 
-        while (screenshot_usage > MAX_CACHE_SIZE) {
-            string? oldest_path = null;
-            time_t oldest_time = 0;
-            while ((name = dir.read_name ()) != null) {
-                string entry = Path.build_filename (screenshot_path, name);
-                Stat fstat = Stat (entry);
-                if (oldest_time == 0 || fstat.st_mtime < oldest_time) {
-                    oldest_time = fstat.st_mtime;
-                    oldest_path = entry;
-                }
-            }
-
-            if (null != oldest_path) {
-                debug ("deleting screenshot at %s to free cache\n", oldest_path);
-
-                var file = File.new_for_path (oldest_path);
-                int size = query_size (oldest_path);
-                if (size == -1) {
-                    return;
-                }
+        FileInfo? info;
 
-                if (delete_file (file, oldest_path) == 0) {
-                    screenshot_usage -= size;
-                } else {
-                    return;
+        // Get a list of the files in the screenshot cache folder
+        try {
+            while ((info = enumerator.next_file (null)) != null) {
+                if (info.get_file_type () == FileType.REGULAR) {
+                    file_list.add (info);
                 }
             }
-        }
-    }
-
-    // Get the combined size of the screenshot cache.
-    private int summarize_screenshot_usage () {
-        Dir dir;
-        if (read_dir (out dir, screenshot_path) == -1) {
-            return -1;
+        } catch (Error e) {
+            warning ("Error while enumerating screenshot cache dir: %s", e.message);
         }
 
-        string? name = null;
-        int size = 0;
+        // Sort the files by ctime (when file metadata was changed, not content)
+        file_list.sort ((a, b) => {
+            uint64 a_time = a.get_attribute_uint64 (GLib.FileAttribute.TIME_CHANGED);
+            uint64 b_time = b.get_attribute_uint64 (GLib.FileAttribute.TIME_CHANGED);
 
-        while ((name = dir.read_name ()) != null) {
-            string entry = Path.build_filename (screenshot_path, name);
-            int file_size = query_size (entry);
-            if (file_size == -1) {
+            if (a_time < b_time) {
```

Comment:
This would do it:
return (int) (b_time - a_time);

Reply:
I'm not so sure...

`(uint64)0 - (uint64)1 == UINT64_MAX`

Then casting UINT64_MAX to integer (int32) would lose precision and you'd get 0, which would not be the correct result.

Reply:
Obviously that's an unrealistic case, but if `b < a` then `b - a` will probably be greater than INT32_MAX which would be truncated to 0 on the cast to `int`, when we'd expect the result to be < 0

Reply:
Yeah you're right, leaving it like this is fine then :)